### PR TITLE
Persist test logs in blobstore

### PR DESCRIPTION
**Pull request by tempoz**
Persist test logs in the blobstore so that they are still visible after they would have fallen
out of the cache.

---

**Version bump**: Minor

### CODE REVIEW

**Comment by tempoz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3717** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3717" target="_blank">Graphite</a> 
    * **PR #3718** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3718" target="_blank">Graphite</a> 
      * **PR #3719** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3719" target="_blank">Graphite</a> 
        * **PR #3720** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3720" target="_blank">Graphite</a> 
          * **PR #3721** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3721" target="_blank">Graphite</a> 
            * **PR #3780** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3780" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3780?utm_source=stack-comment).

File: server/build_event_protocol/accumulator/accumulator.go [hunk 1]
```diff
@@ -127,6 +129,17 @@ func (v *BEValues) AddEvent(event *build_event_stream.BuildEvent) {
...
+		for _, f := range p.TestResult.TestActionOutput {
+			if f.GetName() == "test.log" {
+				if url, err := url.Parse(f.GetUri()); err != nil {
+					log.Warningf("Error parsing uri from TestResult: %s", f.GetUri())
+				} else {
+					v.testLogURI = url
+					v.testLogName = f.GetName()
+				}
+			}
+		}
```
Comment (bduffany) [inline server/build_event_protocol/accumulator/accumulator.go:142]: If there are multiple test results, then I think we'll want the ability to serve multiple log files.

We don't want to add multiple log files to the accumulator though since those are variable in length and in general we want the accumulator to be a flat struct that doesn't depend on the size of the invocation, since it sits in memory while the invocation is in progress.

Instead, we could have the test log endpoint read the invocation and then scan the BES events for the desired test log.

**Review by bduffany — CHANGES_REQUESTED** [review]

File: server/build_event_protocol/build_event_handler/build_event_handler.go [hunk 2]
```diff
@@ -777,6 +777,7 @@ func (e *EventChannel) FinalizeInvocation(iid string) error {
 		invocation,
 		map[string]*url.URL{
 			"timing_profile/" + e.beValues.ProfileName(): e.beValues.ProfileURI(),
+			"test_log/" + e.beValues.TestLogName():       e.beValues.TestLogURI(),
```
Comment (bduffany) [inline server/build_event_protocol/build_event_handler/build_event_handler.go:780]: we'll probably want to scan the BES stream here to get a list of test logs. In `webhookNotifier` (which also runs in the background, similarly to statsRecorder), we are already looking up the full invocation BES event stream. Maybe we could somehow reuse that work? ideally in a way where we aren't keeping the full BES stream buffered in memory for too long.

File: server/build_event_protocol/accumulator/accumulator.go (hunk 1 above)
Comment (tempoz) [reply server/build_event_protocol/accumulator/accumulator.go]: Done!

File: server/build_event_protocol/build_event_handler/build_event_handler.go (hunk 2 above)
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go]: This seemed very complicated to do; I think it is work worth doing, but may be outside the scope of this PR. We probably want to make some sort of BuildEventStreamPostProcess pipeline that we can feed arbitrary functions into and then it can load the BES (ideally in small chunks via streaming) and map all the functions over the events?

File: server/build_event_protocol/build_event_handler/build_event_handler.go [hunk 3]
```diff
@@ -408,35 +416,40 @@ func (r *statsRecorder) handleTask(ctx context.Context, task *recordStatsTask) {
...
-			log.CtxErrorf(ctx, "Failed to commit to blobstore for path %s to persist cache artifact at %s: %s", path, uri.String(), err)
-			w.Close()
-			continue
-		}
-		if err := w.Close(); err != nil {
-			log.CtxErrorf(ctx, "Failed to close blobstore writer for path %s to persist cache artifact at %s: %s", path, uri.String(), err)
+		switch uri.Scheme {
+		case "":
+			switch uri.Path {
+			case "//test_data/":
```
[the remark this answers is not in the document]
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go]: Done.

File: server/build_event_protocol/build_event_handler/build_event_handler.go [hunk 4]
```diff
@@ -464,6 +477,82 @@ func (r *statsRecorder) Stop() {
...
+			if resultURI.Scheme != "bytestream" {
+				continue
+			}
+			if _, ok := testArtifactsToPersist[f.Name]; !ok {
+				continue
+			}
+			target := strings.TrimPrefix(e.BuildEvent.Id.GetTestResult().Label, "//")
+			if i := strings.LastIndex(target, ":"); i != -1 {
+				target = target[:i] + "/" + target[i:]
+			}
```
[the remark this answers is not in the document]
Comment (siggisim) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:539]: I think doing this path manipulation is error prone, and there are all kinds of edge cases that you'd have to account for i.e. test with multiple runs, tests with multiple shards, retries, etc.

You wouldn't have to do a fallback for every request if you used cache paths - you could just take a boolean flag on whether or not to fallback to blobstore, and we could set that in the UI on places where there might be a fallback (i.e. test logs and timing profiles for now).

[the remark this answers is not in the document]
Comment (bduffany) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:539]: Not sure if this is what you meant @siggisim but instead of using the path at all, we could use the SHA256 of the test artifact as the blobstore path. e.g. if the file name is bytestream://remote.buildbuddy.io/blob/abc123/456 then we could store it in blobstore under `{invocation_id}/artifacts/sha256/abc123` -- would that work?

[the remark this answers is not in the document]
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go]: switched to cache-style paths.

File: server/build_event_protocol/build_event_handler/build_event_handler.go [hunk 5]
```diff
@@ -93,6 +94,13 @@ var (
 			"local cache stats to the backing storage, before finalizing stats in the DB.")
 )
 
+var cacheArtifactsBucket = path.Join("artifacts", "cache")
```
Comment (bduffany) [inline server/build_event_protocol/build_event_handler/build_event_handler.go:97]: nit: s/Bucket/BlobstorePath/ (since bucket has a different meaning GCS)

File: server/buildbuddy_server/buildbuddy_server.go [hunk 6]
```diff
@@ -1187,16 +1188,10 @@ func (s *BuildBuddyServer) serveUsingParams(w http.ResponseWriter, r *http.Reque
...
-		// a fallback is specified.
-		if fallback, parseErr := url.Parse(params.Get("with_fallback")); parseErr != nil {
-			http.Error(w, parseErr.Error(), http.StatusBadRequest)
-		} else if fallback.RawQuery == "" {
-			http.Error(w, fmt.Sprintf("Fallback without query is invalid: %v", fallback), http.StatusBadRequest)
-		} else {
-			q := fallback.Query()
-			q.Del("with_fallback") // Only fall back once.
-			s.serveUsingParams(w, r, q)
+		// Attempt to fall back gracefully if the specified file is not found
```
Comment (bduffany) [inline server/buildbuddy_server/buildbuddy_server.go:1191]: I think we should only do this fallback in the "bytestream_url" case above, since it looks like in the "artifact" case it's just a retry and not a fallback, if I'm reading it right?

**Review by bduffany — APPROVED** [review]
